### ISSUE TITLE
[java] PMD could enforce non-public methods for Junit5 / Jupiter test methods

### ISSUE THREAD

**Issue by ajeans** [pmd/pmd#3239]
**Proposed Rule Name:** JUnit5TestShouldBePackagePrivate

**Proposed Category:** Best Practices

**Description:**
JUnit 5 tests should be package private at the class level and for each method that uses `@Test`, `@RepeatedTest`,
`@TestFactory`, `@TestTemplate` or `@ParameterizedTest`.

Contrary to JUnit4 tests that required public visibility to be run by the engine, JUnit5 tests can also be run
if they're package-private. Marking them as such is a good practice to limit their visibility.

**Code Sample:** 

```java
class MyTest { // not public, that's fine
    @Test
    public void testBad() { } // should not have a public modifier
    @Test
    protected void testAlsoBad() { } // should not have a protected modifier
    @Test
    void testGood() { } // package private as expected
}
```

**Possible Properties:**

*   There is no necessity for a property yet.

-----

<details>
  <summary>Original feature request</summary>

<p>

**Is your feature request related to a problem? Please describe.**

We work with Junit 5 for our tests and PMD does not warn us when declare test methods as `public`. This is not needed anymore, so we would like PMD to assert that methods are not public nor private.
See https://junit.org/junit5/docs/current/user-guide/#writing-tests-classes-and-methods

Slightly related to https://github.com/pmd/pmd/issues/428 which was PMD incorrectly requiring the public modifier, we would like to go further and have PMD require that the modifier should not be public.

**Describe the solution you'd like**
An additional rule JUnit5TestNoPublicOrPrivateModifier to error out on methods marked as `@Test` and with a public or private modifier.

**Describe alternatives you've considered**
Adding Sonar to our CI, which reports this as an issue

**Additional context**
Some other ideas for Junit 5 are in https://github.com/pmd/pmd/issues/2616

</p>
</details>

**Comment by ajeans** [comment]
Looking into this locally, it seems to be doable with another XPath rule like

```
//ClassOrInterfaceDeclaration[
       matches(@SimpleName, $testClassPattern)]

    /ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration
        [MethodDeclaration[@Public=true() or @Private=true()]]
        [Annotation//Name[
            pmd-java:typeIs('org.junit.jupiter.api.Test') or pmd-java:typeIs('org.junit.jupiter.api.RepeatedTest')
            or pmd-java:typeIs('org.junit.jupiter.api.TestFactory') or pmd-java:typeIs('org.junit.jupiter.api.TestTemplate')
            or pmd-java:typeIs('org.junit.jupiter.params.ParameterizedTest')
        ]]
```

**Comment by ajeans** [comment]
I can do a MR if there is interest...

**Comment by adangel** [comment]
Thanks for the suggestion!

I'm currently thinking about how this new rule would fit into the existing rules we have for (J)Unit testing. Some rules have "JUnit4" in their name, but actually consider Junit 3-5 (e.g. [JUnit4TestShouldUseTestAnnotation](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#junit4testshouldusetestannotation)), which means, we probably should rename the rule at some point.

This new rule would do two things:
* It flags test methods, that are private: This is an error - since it won't work.
* It flags test methods, that are public: This is a matter of taste. The test still works, but this is not required anymore with JUnit5. But it is also not a must. PMD could declare it as a "best practice" though.

Looking at the sonar rules, I find these:
* https://rules.sonarsource.com/java/tag/tests/RSPEC-5810 -> this is the error case
* https://rules.sonarsource.com/java/tag/tests/RSPEC-5786 -> which is exactly the second case (although a bit extended, because it also flags "protected" and only allows package-private). I was interested for the rationale, but it just says "... to use the default package visibility, which improves readability of code".

If we would do a single rule, I could imagine a name like "CorrectTestMethodDeclaration" - but since it finds an error and a best-practice-violation, it is difficult to find a single category...

So maybe we should really have two rules:
* JUnit5TestNoPrivateModifier in category "error prone"
* JUnit5TestShouldBePackagePrivate in category "best practice" or "code style"

I'm still unsure about the names containing "JUnit5" explicitly, but maybe these rules really only apply to JUnit5 and no other test frameworks. The first rule "JUnit5TestNoPrivateModifier" would probably also apply to JUnit4, so the name could also be "JUnitTestWrongModifier" (for JUnit4, private, protected and no modifier would be flagged, for JUnit5 only private modifier would be flagged). The second rule would only apply for JUnit5.

**Comment by ajeans** [comment]
> 
    JUnit5TestNoPrivateModifier in category "error prone"
    JUnit5TestShouldBePackagePrivate in category "best practice" or "code style"

Sounds good

Thinking further about jupiter support, I noticed that I didn't mention public jupiter classes, which I also not recommended and covered by https://rules.sonarsource.com/java/RSPEC-5786 (which covers both public methods and public classes).

I wrote a separate xpath rule for the public class case (adding non abstract) and ended up with 

```
//ClassOrInterfaceDeclaration[
       matches(@SimpleName, $testClassPattern) and
       @Abstract=false() and
       @Public=true()
       ]

    /ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration
        [MethodDeclaration]
        [Annotation//Name[
            pmd-java:typeIs('org.junit.jupiter.api.Test') or pmd-java:typeIs('org.junit.jupiter.api.RepeatedTest')
            or pmd-java:typeIs('org.junit.jupiter.api.TestFactory') or pmd-java:typeIs('org.junit.jupiter.api.TestTemplate')
            or pmd-java:typeIs('org.junit.jupiter.params.ParameterizedTest')
        ]]
]]>
``` 

However I don't see how to implement the two ideas (public class and public method) into a single XPath rule for `JUnit5TestShouldBePackagePrivate`

So ``JUnit5TestMethodShouldBePackagePrivate` and `JUnit5TestClassShouldBePackagePrivate ?

**Comment by oowekyala** [comment]
> However I don't see how to implement the two ideas (public class and public method) into a single XPath rule for JUnit5TestShouldBePackagePrivate

You can join the two xpath expressions with the union operator: `|`

**Comment by ajeans** [comment]
:arrow_up: I pushed a simple MR for `JUnit5TestShouldBePackagePrivate`, I am interested in seeing this upstream :)

**Comment by MPoorter** [comment]
@ajeans @adangel Why is protected considered wrong here? We have certain tests defined in abstract classes that get used in subclasses that are not in the same package. This rule now causes issues, can we allow for protected modifiers on the test methods?

**Comment by ajeans** [comment]
Hello @MPoorter 

Can you provide a short example that shows the problem?

I understand that you are inheriting protected methods annotated with `@org.junit.jupiter.api.Test` from the parent and running them directly. Is that correct?

What happens if you convert the parent class to an interface with `default` methods (see https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-interfaces-and-default-methods)?

Thanks.

**Comment by MPoorter** [comment]
Hi @ajeans 

Our abstract class looks like this

```
public abstract class AbstractFeignClientTest {

    @Test
    void testHystrixEmptyResponse() {
        doFaultTest(EMPTY_RESPONSE);
    }

    @Test
    void testHystrixConnectionReset() {
        doFaultTest(CONNECTION_RESET_BY_PEER);
    }

    @Test
    @SuppressWarnings("PMD.JUnit5TestShouldBePackagePrivate")
    protected void testHystrixMalformedResponse() {
        doFaultTest(MALFORMED_RESPONSE_CHUNK);
    }

    @Test
    void testHystrixRandomData() {
        doFaultTest(RANDOM_DATA_THEN_CLOSE);
    }

    protected abstract void executeFault(Fault fault);

    protected abstract ResponseDefinitionBuilder success();

    protected abstract MappingBuilder getRequest();

    private void doFaultTest(final Fault fault) {
        stubFor(getRequest().willReturn(success().withFault(fault)));
        executeFault(fault);
    }
}
```

As you can see, most of the tests defined in the abstract class are package private and don't cause any issues. However, occasionally we cannot rely on the default implementation for the `testHystrixMalformedResponse` and we override it in a subclass that is not in the same package. Therefore, we need to specify it's protected so that subclasses have access to this method.

I'm not sure whether the interface solution would work for us, I would need to look into it further but we do have Spring bean injection in our abstract class (not shown here), which makes me wonder whether that would work for us.

I also wonder whether the interface is the correct solution, as I understand that any default methods in an interface are by nature public. If this rule is set up to prevent public methods for Junit5 tests, then default test methods in interfaces should break that rule, even if there is no access modifier specified.

The way I understood the benefit of this rule is that with Junit5, we no longer need to say that test methods are public (unlike in Junit4). However, I don't understand why protected also needs to be taken up in this rule?

Happy to hear your thoughts on this,
Thanks,
Matt

**Comment by ajeans** [comment]
Thanks @MPoorter 

I am on the fence on this.

I can see how useful it is to use abstract classes and inheritance to share tests. And the change would be trivial.

But your use case:
* is way less common than someone using `protected` by mistake with a simple test layout (one single class)
* would also warn in Sonar (see https://rules.sonarsource.com/java/tag/tests/RSPEC-5786) 
* could be done differently by wrapping the common behaviour (inheritance vs composition)
* can be safely ignored with the targeted `@SuppressWarnings("PMD.JUnit5TestShouldBePackagePrivate")` you are already using

Also, the rule is named `JUnit5TestShouldBePackagePrivate`. If we stop complaining about protected test methods, we might as well rename it to convey that we only complain about `public` modifiers on tests.

What do you think @adangel?
